### PR TITLE
Enable Syslog log handler

### LINKED ISSUE #27544: Expose quarkus syslog logging now GELF is being deprecated from Keycloak
### Description

GELF logging will be removed from Keycloak because the upstream library has been deprecated: 
https://github.com/keycloak/keycloak/issues/27365

This leaves Keycloak with zero built-in mechanism for remote logging. Is it possible to expose the built-in syslog logging mechanism of Quarkus to Keycloak? This does not add extra dependencies because Keycloak is nowadays built on top of Quarkus.

Pre-quarkus builds of Keycloak could log to syslog, so people that still needed a remote logger that could handle volume moved to the newly GELF logging. Restoring syslog functionality with easy the transition for these kind of setups.

Thank you.

Edit: Quarkus documentation syslog logging: https://quarkus.io/guides/logging#quarkus-logging-json_quarkus-log-syslog-json-syslog-logging

### Discussion

_No response_

### Motivation

- Keycloak supported Syslog logging.
- While moving to Quarkus, Keycloak lost this functionality while introducing GELF logging as an alternative.
- Keycloak will remove GELF and will break the logging setup of people that needs remote logging for high volumes.

### Details

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #27544

Providing Syslog log handler was actually low-hanging fruit, as Quarkus supports it by default as one of the possible log handlers.

### CODE REVIEW

File: docs/guides/server/logging.adoc
Comment (ahus1) [inline docs/guides/server/logging.adoc:244]: Nitpick: Don't include "mabartos" as the hostname, and possily pick a different log message that will be more timeless than "Keycloak 999.0.0-SNAPSHOT on JVM (powered by Quarkus 3.8.3)"

**Review by ahus1 — COMMENTED** [review]
Thank you for this PR, this is great to see and it will be a good improvement for those running Keycloak in bare-metal environments, which is still a significant share. 

Commenting here only as I want to leave the final approval for the CND team.

File: docs/guides/server/logging.adoc
Comment (mabartos) [reply docs/guides/server/logging.adoc:244]: @ahus1 Ooops, forgot to remove it. I thought keeping the "convention" in the guide would be good. But I agree that we can find a better message. Thanks

Comment (ahus1) [reply docs/guides/server/logging.adoc]: Oh, I now see the other entries which state "Keycloak 18.0.0-SNAPSHOT". Changing this might then be out-of-scope for this PR (or not if decide against it).

**Comment by mabartos** [comment]
@shawkins jfyi - I tried to use your approach around the HelpCommandDistTest, but it seems it does not work as expected. I'll try to analyze it more. Not sure what's happening there yet. 

https://github.com/keycloak/keycloak/actions/runs/8568547697/job/23482814768?pr=28462

**Comment by shawkins** [comment]
> Not sure what's happening there yet.

So you ran it with the env variable and it passed, you checked in the expectation changes, and then it failed on ci? Seems very odd.

**Comment by mabartos** [comment]
Not sure what happened there, but it works now... I probably did sth wrong :))

**Comment by mabartos** [comment]
@keycloak/cloud-native Could you please check it? Thanks

**Review by shawkins — DISMISSED** [review]
LGTM, thanks for getting this in @mabartos it should make the gelf deprecation much easier.

Should there be a note in the gelf docs pointing to syslog or vice versa so that users are more directly aware of the alternative?

**Review by Pepo48 — DISMISSED** [review]
LGTM, nice work, thanks @mabartos.

**Review by shawkins — APPROVED** [review]

**Comment by mabartos** [comment]
@shawkins @Pepo48 Thanks guys, for the review. 

I've just added a brief message to the GELF deprecation note.

**Review by Pepo48 — APPROVED** [review]

File: docs/guides/server/logging.adoc
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:204]: One small typo.
```suggestion
To configure the endpoint (_host:port_) of your centralized logging system, enter the following command and substitute the values with your specific values:
```

**Review by vmuzikar — APPROVED** [review]
LGTM, thank you!

File: docs/guides/server/logging.adoc
Comment (mabartos) [reply docs/guides/server/logging.adoc:204]: Oops, auto-merge was on. Can be done in a different PR.

**Comment by LiamLamb** [comment]
Hi all,

I just wanted to inform you of a small change in quarkus around syslog configs: they now support providing the `maxLength` property to the underlying JBoss library.

I am not sure what version of quarkus this will be released in as it's in [main](https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/logging/SyslogConfig.java) but not in their latest [3.9.3 release](https://github.com/quarkusio/quarkus/blob/3.9.3/core/runtime/src/main/java/io/quarkus/runtime/logging/SyslogConfig.java).

See [an issue](https://github.com/quarkusio/quarkus/issues/39944) I opened with quarkus recently and the associated PR [here](https://github.com/quarkusio/quarkus/pull/39958).

**Comment by mabartos** [comment]
@LiamLamb Thanks for the info. We'd be glad to gather some feedback on the recent Syslog integration into Keycloak as providing other necessary options, or just some nice-to-have ones,...

You can create an RFE for providing the option even for Keycloak, and we can discuss there whether it's suitable to allow it through the Keycloak props, or just use the quarkus.properties directly.

**Comment by ahus1** [comment]
@LiamLamb - thank you for this PR to the Quarkus project. 

Some background information on Quarkus versions in Keycloak: Keycloak stays on the LTS verison of Quarkus until there is a new LTS release which happens about every 6 (?) months and will be scheduled possibly together with KC26.  

The current LTS version is Quarkus 3.8. If your change is small, and proves to be stable in Quarkus 3.9 or 3.10, you might ask to get it possibly backported to a Quarkus 3.8 patch release. Still this is up for the Quarkus team to decide.

**Comment by LiamLamb** [comment]
@ahus1 @mabartos  Thanks for the speedy replies!
That makes sense - I've pinged the folks over at quarkus and they will consider backporting this into their next 3.8.x release 👍

**Comment by trixpan** [comment]
@mabartos I may be late to the party but I have a question:

From my reading it seems like this PR sort of exposes a configuration of  syslog via keycloak but leaves out of the PR a number of pretty important aspects of the configuration:

* TCP vs UDP:
No serious security solution should be using UDP syslog anymore.  A key reason one uses remote logging is to aggregate logs and in the particular in the case of KC, audit trails captured via event handling. Now imagine telling your customers: Sorry but we don't know if the access was granted or not because the logs were lost between source and destination (since UDP simply leads to data being discarded silently)

* Async and Queueing:
Given UDP logging should not be used... TCP syslog is a required but far from being a turn-key solution. Queueing and Async / Sync trade-off must be taken care of. JBOS LogManager via Quarkus allows one to semi handle this (tbh it doesn't but that is beside the point), however, these configs don't seem to be exposed.

TBH I find Quarkus decision to include GELF puzzling. The project still seems to suffer from a difference of opinions between  the meaning of their clame to be `Cloud Native: Embraces [12 factor architecture](https://12factor.net/) in environments like Kubernetes.` means given the [same linked site](https://12factor.net/logs) states:

> A twelve-factor app never concerns itself with routing or storage of its output stream. It should not attempt to write to or manage logfiles. Instead, each running process writes its event stream, unbuffered, to stdout. During local development, the developer will view this stream in the foreground of their terminal to observe the app’s behavior.

Not that I agree with the 12 factors view either (raw, unformatted stdout logging is terrible, specially in JAVA), but they have a point in regards to routing: Don't bother trying to solve that in your app, leave that to the execution environment to handle.

All we should focus on is to create a production ready ability to capture logs (and I mean all logs as Wildfly used to offer) using a particular handler that can be configured via properties. The rest should be handled by the execution environment.

**Comment by mabartos** [comment]
@trixpan Thanks for the detailed info!

> TCP vs UDP:

With Keycloak, it is possible to change the syslog protocol via the property `log-syslog-protocol`[1], where the default protocol is TCP. However, even when I'm aligned with your opinion around UDP and logs, it is still possible to change it if a user wants it - and to be aligned with the Quarkus approach. I'm not sure what we can do more in this area.

> Async and Queueing:

It is true that we do not provide the Syslog Async properties. If there's a need from the community, we can expose them. I'll analyze if there are some tweaks around async support for Syslog in Keycloak, but we should be good. I create a follow-up enhancement for it. 

>All we should focus on is to create a production ready ability to capture logs (and I mean all logs as Wildfly used to offer) using a particular handler that can be configured via properties. The rest should be handled by the execution environment.

+1


[1] Syslog protocol option https://github.com/keycloak/keycloak/blob/1e3837421e9cea41aa91ef99800af1f82f91089b/quarkus/config-api/src/main/java/org/keycloak/config/LoggingOptions.java#L151

**Comment by mabartos** [comment]
> There will always be a missing flag

@trixpan Yes, it will :)) These Quarkus options, which are not mapped to Keycloak options, are not officially supported by Keycloak, and users use them at their own risk. 

> Furthermore, Quarkus already provides those flags, doesn't it? Does keycloak work with them without any further code on our side? Last time I checked it did, and hence my comment: why botter adding the flags? More code to test, more edge cases to handle.

As mentioned here[1], it is possible to specify Quarkus options in `quarkus.properties` file if users' deployments require it, but as mentioned, it's not officially supported. I know that some of Keycloak properties are just directly mapped to the Quarkus ones without any additional logic around it, but at least, we have some control over them. 

Some of these options just need to be exposed to the Keycloak user and create the whole component context even though they're not transformed in any way. Like HTTP server options - the relevant options need to be tied together.

I know it's not ideal to "duplicate" these options, but thanks to that, we're creating a proper UI for Keycloak users.

> Instead I posit we should focus on instructing people on how to configure the relevant quarkus flags via environment variables or properties file.

The problem here is where these instructions should be posted. It'd be kinda weird to put them into the official documentation and state it's actually not supported because you're using the raw Quarkus properties. 

If we support specifying the Quarkus properties, users would be able to change already mapped quarkus properties or touch properties, which could potentially redirect Keycloak to some "inconsistent" state - and users could say that's the supported way (harder than thought :)).

> IMNSHO, the less we do outside quarkus, the better.

Yes, that's true, but for the Syslog async, it might make sense to support it in order to improve the overall UX around Syslog as it might be quite required enhancement to the Syslog capabilities. 

@trixpan Or do you think exposing and supporting Syslog async options might bring more drawbacks than benefits? And rather instruct users (somehow) how to achieve it via the Quarkus props?

[1] https://www.keycloak.org/server/configuration#_format_for_raw_quarkus_properties

**Pull request merged**
